File: asterix-rs/tests/integration_test.rs
//! Integration tests for ASTERIX Rust bindings
//!
//! Tests parsing of real ASTERIX data files and validates against expected outputs.

use asterix::{describe, parse, parse_with_offset, AsterixError, ParseOptions};
use std::fs;
use std::path::PathBuf;

/// Helper function to get path to sample data
fn sample_data_path(filename: &str) -> PathBuf {
    let mut path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    path.pop(); // Go up from asterix-rs/
    path.push("install/sample_data");
    path.push(filename);
    path
}

#[test]
fn test_parse_cat048_raw() {
    let path = sample_data_path("cat048.raw");
    let data = fs::read(&path).expect("Failed to read cat048.raw");

    let options = ParseOptions {
        verbose: false,
        filter_category: None,
        max_records: None,
    };

    let records = parse(&data, options).expect("Failed to parse cat048.raw");

    // Validate results
    assert!(!records.is_empty(), "Expected at least one record");

    // Check first record
    let first = &records[0];
    assert_eq!(first.category, 48, "Expected category 48");
    assert!(first.length > 0, "Record length should be positive");
    assert!(!first.hex_data.is_empty(), "Hex data should not be empty");

    // Validate items exist
    assert!(!first.items.is_empty(), "Record should contain data items");

    println!("✓ Parsed {} CAT048 record(s)", records.len());
}

#[test]
fn test_parse_cat062_cat065_raw() {
    let path = sample_data_path("cat062cat065.raw");
    let data = fs::read(&path).expect("Failed to read cat062cat065.raw");

    let records = parse(&data, ParseOptions::default()).expect("Failed to parse");

    assert!(!records.is_empty(), "Expected records");

    // Count categories
    let cat62_count = records.iter().filter(|r| r.category == 62).count();
    let cat65_count = records.iter().filter(|r| r.category == 65).count();

    println!(
        "✓ Found {} CAT062 and {} CAT065 records",
        cat62_count, cat65_count
    );
    assert!(
        cat62_count > 0 || cat65_count > 0,
        "Expected CAT062 or CAT065 records"
    );
}

#[test]
fn test_parse_pcap_format() {
    let path = sample_data_path("cat_062_065.pcap");
    let data = fs::read(&path).expect("Failed to read PCAP file");

    let options = ParseOptions {
        verbose: true,
        filter_category: None,
        max_records: None,
    };

    let records = parse(&data, options).expect("Failed to parse PCAP");

    assert!(!records.is_empty(), "Expected records from PCAP");

    // Verify PCAP contains multiple categories
    let categories: std::collections::HashSet<_> = records.iter().map(|r| r.category).collect();

    println!("✓ Parsed PCAP with categories: {:?}", categories);
    assert!(categories.len() >= 1, "Expected at least one category");
}

#[test]
fn test_parse_cat034_048_pcap() {
    let path = sample_data_path("cat_034_048.pcap");
    let data = fs::read(&path).expect("Failed to read cat_034_048.pcap");

    let records = parse(&data, ParseOptions::default()).expect("Failed to parse");

    assert!(!records.is_empty(), "Expected records");

    let cat34_count = records.iter().filter(|r| r.category == 34).count();
    let cat48_count = records.iter().filter(|r| r.category == 48).count();

    println!("✓ CAT034: {}, CAT048: {}", cat34_count, cat48_count);
    assert!(
        cat34_count > 0 || cat48_count > 0,
        "Expected CAT034 or CAT048"
    );
}

#[test]
fn test_parse_multicast_pcap() {
    let path = sample_data_path("asterix.pcap");
    let data = fs::read(&path).expect("Failed to read asterix.pcap");

    let records = parse(&data, ParseOptions::default()).expect("Failed to parse");

    println!("✓ Parsed {} records from multicast capture", records.len());
    assert!(!records.is_empty(), "Expected records from multicast PCAP");

    // Check record structure
    for (i, record) in records.iter().take(5).enumerate() {
        println!(
            "  Record {}: CAT{:03} ({} bytes)",
            i, record.category, record.length
        );
        assert!(record.category > 0, "Invalid category");
        assert!(record.length > 0, "Invalid length");
    }
}

#[test]
fn test_parse_gps_format() {
    let path = sample_data_path("parsegps.gps");
    let data = fs::read(&path).expect("Failed to read GPS file");

    // GPS format parsing (may fail if not supported in Rust yet)
    let result = parse(&data, ParseOptions::default());

    match result {
        Ok(records) => {
            println!("✓ Parsed {} GPS records", records.len());
            assert!(!records.is_empty(), "Expected GPS records");
        }
        Err(e) => {
            println!("⚠ GPS format not yet supported: {:?}", e);
            // This is acceptable if GPS format isn't implemented yet
        }
    }
}

#[test]
fn test_error_handling_invalid_data() {
    let invalid_data = b"This is not ASTERIX data at all!";

    let result = parse(invalid_data, ParseOptions::default());

    assert!(result.is_err(), "Expected parse error for invalid data");

    match result {
        Err(AsterixError::ParseError { offset, message }) => {
            println!(
                "✓ Correctly rejected invalid data at offset {} ({})",
                offset, message
            );
        }
        Err(e) => {
            println!("✓ Rejected with error: {:?}", e);
        }
        Ok(_) => panic!("Should not successfully parse invalid data"),
    }
}

#[test]
fn test_error_handling_empty_data() {
    let empty_data = b"";

    let result = parse(empty_data, ParseOptions::default());

    assert!(result.is_err(), "Expected error for empty data");
    println!("✓ Correctly rejected empty input");
}

#[test]
fn test_error_handling_truncated_data() {
    // Create truncated ASTERIX data (incomplete header)
    let truncated = b"\x30\x00"; // CAT048, but missing length bytes

    let result = parse(truncated, ParseOptions::default());

    assert!(result.is_err(), "Expected error for truncated data");

    match result {
        Err(AsterixError::UnexpectedEOF) => {
            println!("✓ Detected unexpected EOF in truncated data");
        }
        Err(e) => {
            println!("✓ Rejected truncated data: {:?}", e);
        }
        Ok(_) => panic!("Should not parse truncated data"),
    }
}

#[test]
fn test_incremental_parsing() {
    let path = sample_data_path("cat_034_048.pcap");
    let data = fs::read(&path).expect("Failed to read PCAP");

    let mut offset = 0;
    let mut total_records = 0;
    let mut iterations = 0;

    loop {
        let result = parse_with_offset(
            &data,
            offset,
            10, // Parse 10 blocks at a time
            ParseOptions::default(),
        );

        match result {
            Ok(parse_result) => {
                total_records += parse_result.records.len();
                offset = parse_result.bytes_consumed;
                iterations += 1;

                println!(
                    "  Iteration {}: parsed {} records, consumed {} bytes",
                    iterations,
                    parse_result.records.len(),
                    parse_result.bytes_consumed
                );

                if parse_result.remaining_blocks == 0 {
                    break;
                }

                // Safety check to prevent infinite loops
                if iterations > 1000 {
                    panic!("Too many iterations in incremental parsing");
                }
            }
            Err(e) => {
                println!("⚠ Incremental parsing stopped: {:?}", e);
                break;
            }
        }
    }

    println!(
        "✓ Incremental parsing completed: {} total records in {} iterations",
        total_records, iterations
    );
    assert!(total_records > 0, "Expected at least some records");
}

#[test]
fn test_parse_with_category_filter() {
    let path = sample_data_path("cat_062_065.pcap");
    let data = fs::read(&path).expect("Failed to read PCAP");

    // Parse with CAT062 filter
    let options = ParseOptions {
        verbose: false,
        filter_category: Some(62),
        max_records: None,
    };

    let records = parse(&data, options).expect("Failed to parse with filter");

    // Verify all records are CAT062
    for record in &records {
        assert_eq!(record.category, 62, "Expected only CAT062 records");
    }

    println!(
        "✓ Category filter working: {} CAT062 records",
        records.len()
    );
}

#[test]
fn test_parse_with_max_records() {
    let path = sample_data_path("cat_034_048.pcap");
    let data = fs::read(&path).expect("Failed to read PCAP");

    let options = ParseOptions {
        verbose: false,
        filter_category: None,
        max_records: Some(5),
    };

    let records = parse(&data, options).expect("Failed to parse with limit");

    assert!(records.len() <= 5, "Should not exceed max_records limit");
    println!("✓ Max records limit working: got {} records", records.len());
}

#[test]
fn test_describe_category() {
    // Describe CAT048
    let result = describe(48, None, None, None);

    assert!(result.is_ok(), "Failed to describe CAT048");

    let description = result.unwrap();
    assert!(!description.is_empty(), "Description should not be empty");

    println!(
        "✓ CAT048 description: {}",
        description.chars().take(100).collect::<String>()
    );
}

#[test]
fn test_describe_item() {
    // Describe CAT048 item 010 (Data Source Identifier)
    let result = describe(48, Some("010"), None, None);

    match result {
        Ok(description) => {
            println!("✓ CAT048/010 description: {}", description);
            assert!(!description.is_empty());
        }
        Err(e) => {
            println!("⚠ Describe item failed: {:?}", e);
            // May not be implemented yet
        }
    }
}

#[test]
fn test_describe_invalid_category() {
    // Try to describe non-existent category
    let result = describe(255, None, None, None);

    assert!(result.is_err(), "Should fail for invalid category");

    match result {
        Err(AsterixError::InvalidCategory { category, .. }) => {
            assert_eq!(category, 255);
            println!("✓ Correctly rejected invalid category {}", category);
        }
        Err(e) => {
            println!("✓ Rejected with error: {:?}", e);
        }
        Ok(_) => panic!("Should not succeed for invalid category"),
    }
}

#[test]
fn test_record_structure() {
    let path = sample_data_path("cat048.raw");
    let data = fs::read(&path).expect("Failed to read file");

    let records = parse(&data, ParseOptions::default()).expect("Failed to parse");
    let record = &records[0];

    // Validate record structure
    assert!(record.category > 0 && record.category < 256);
    assert!(record.length >= 3); // Minimum ASTERIX record size
    assert!(record.timestamp_ms >= 0);
    assert!(!record.hex_data.is_empty());

    // Validate hex data format (should be valid hex string)
    assert!(record
        .hex_data
        .chars()
        .all(|c| c.is_ascii_hexdigit() || c.is_whitespace()));

    println!("✓ Record structure validated");
    println!("  Category: {}", record.category);
    println!("  Length: {}", record.length);
    println!("  Timestamp: {} ms", record.timestamp_ms);
    println!("  Items: {}", record.items.len());
}

#[test]
fn test_data_item_structure() {
    let path = sample_data_path("cat048.raw");
    let data = fs::read(&path).expect("Failed to read file");

    let records = parse(&data, ParseOptions::default()).expect("Failed to parse");
    let record = &records[0];

    // Check that we have data items
    assert!(!record.items.is_empty(), "Record should have data items");

    // Examine first item
    if let Some((item_name, item)) = record.items.iter().next() {
        println!("✓ First data item: {}", item_name);

        if let Some(desc) = &item.description {
            println!("  Description: {}", desc);
        }

        println!("  Fields: {}", item.fields.len());

        // Check fields
        for (field_name, value) in &item.fields {
            println!("    {} = {:?}", field_name, value);
        }
    }
}

#[test]
fn test_parsed_value_types() {
    use asterix::ParsedValue;

    let path = sample_data_path("cat048.raw");
    let data = fs::read(&path).expect("Failed to read file");

    let records = parse(&data, ParseOptions::default()).expect("Failed to parse");
    let record = &records[0];

    // Collect statistics on value types
    let mut int_count = 0;
    let mut float_count = 0;
    let mut string_count = 0;
    let mut bool_count = 0;
    let mut bytes_count = 0;

    for item in record.items.values() {
        for value in item.fields.values() {
            match value {
                ParsedValue::Integer(_) => int_count += 1,
                ParsedValue::Float(_) => float_count += 1,
                ParsedValue::String(_) => string_count += 1,
                ParsedValue::Boolean(_) => bool_count += 1,
                ParsedValue::Bytes(_) => bytes_count += 1,
                ParsedValue::Nested(_) => {} // Count as parent type
            }
        }
    }

    println!("✓ Value type distribution:");
    println!("  Integers: {}", int_count);
    println!("  Floats: {}", float_count);
    println!("  Strings: {}", string_count);
    println!("  Booleans: {}", bool_count);
    println!("  Bytes: {}", bytes_count);

    let total = int_count + float_count + string_count + bool_count + bytes_count;
    assert!(total > 0, "Should have parsed some values");
}

#[test]
#[cfg(feature = "serde")]
fn test_serialization_to_json() {
    use serde_json;

    let path = sample_data_path("cat048.raw");
    let data = fs::read(&path).expect("Failed to read file");

    let records = parse(&data, ParseOptions::default()).expect("Failed to parse");

    // Serialize to JSON
    let json = serde_json::to_string_pretty(&records).expect("Failed to serialize");

    println!("✓ Serialized to JSON ({} bytes)", json.len());
    assert!(!json.is_empty());

    // Verify JSON is valid by deserializing
    let _deserialized: Vec<asterix::AsterixRecord> =
        serde_json::from_str(&json).expect("Failed to deserialize");

    println!("✓ JSON round-trip successful");
}

#[test]
fn test_concurrent_parsing() {
    use std::sync::Arc;
    use std::thread;

    let path = sample_data_path("cat048.raw");
    let data = Arc::new(fs::read(&path).expect("Failed to read file"));

    let mut handles = vec![];

    // Spawn multiple threads parsing the same data
    for i in 0..4 {
        let data_clone = Arc::clone(&data);
        let handle = thread::spawn(move || {
            let records =
                parse(&data_clone, ParseOptions::default()).expect("Failed to parse in thread");
            (i, records.len())
        });
        handles.push(handle);
    }

    // Collect results
    let mut results = vec![];
    for handle in handles {
        results.push(handle.join().expect("Thread panicked"));
    }

    // Verify all threads got same results
    let first_count = results[0].1;
    for (thread_id, count) in &results {
        assert_eq!(
            *count, first_count,
            "Thread {} got different count",
            thread_id
        );
    }

    println!(
        "✓ Concurrent parsing successful: {} threads, {} records each",
        results.len(),
        first_count
    );
}

#[test]
fn test_memory_safety_large_file() {
    // Test with largest available sample file
    let path = sample_data_path("cat_034_048.pcap");
    let data = fs::read(&path).expect("Failed to read file");

    println!("Testing with {} byte file", data.len());

    // Parse multiple times to check for memory leaks/corruption
    for iteration in 0..10 {
        let records = parse(&data, ParseOptions::default()).expect("Failed to parse in iteration");

        if iteration == 0 {
            println!("  Parsed {} records", records.len());
        }

        // Verify records are still valid
        for record in &records {
            assert!(record.category > 0);
            assert!(record.length > 0);
        }
    }

    println!("✓ Memory safety check passed (10 iterations)");
}

#[test]
fn test_verbose_mode() {
    let path = sample_data_path("cat048.raw");
    let data = fs::read(&path).expect("Failed to read file");

    // Parse with verbose mode
    let options = ParseOptions {
        verbose: true,
        filter_category: None,
        max_records: None,
    };

    let records = parse(&data, options).expect("Failed to parse with verbose");

    println!(
        "✓ Verbose mode parsing successful: {} records",
        records.len()
    );
    assert!(!records.is_empty());
}

#[test]
fn test_compare_with_python_output() {
    // This test compares Rust output with Python module output
    // Python reference implementation should be available

    let path = sample_data_path("cat048.raw");
    let data = fs::read(&path).expect("Failed to read file");

    let rust_records = parse(&data, ParseOptions::default()).expect("Failed to parse in Rust");

    // TODO: Run Python parser and compare outputs
    // For now, just validate Rust output structure
    println!("✓ Rust parsed {} records", rust_records.len());

    // Validate structure matches Python expectations
    for record in &rust_records {
        assert!(record.category > 0);
        assert!(record.length > 0);
        assert!(!record.items.is_empty());
    }

    println!("✓ Output structure matches Python expectations");
}

#[cfg(test)]
mod benchmarks {
    use super::*;
    use std::time::Instant;

    #[test]
    fn bench_parse_performance() {
        let path = sample_data_path("cat_034_048.pcap");
        let data = fs::read(&path).expect("Failed to read file");

        // Warmup
        let _ = parse(&data, ParseOptions::default()).expect("Warmup failed");

        // Benchmark
        let iterations = 100;
        let start = Instant::now();

        for _ in 0..iterations {
            let _ = parse(&data, ParseOptions::default()).expect("Parse failed");
        }

        let elapsed = start.elapsed();
        let avg_ms = elapsed.as_millis() as f64 / iterations as f64;
        let throughput_mbps =
            (data.len() as f64 * iterations as f64) / (elapsed.as_secs_f64() * 1_000_000.0);

        println!("✓ Performance test ({} iterations):", iterations);
        println!("  Average: {:.2} ms/parse", avg_ms);
        println!("  Throughput: {:.2} MB/s", throughput_mbps);
        println!("  File size: {} bytes", data.len());

        // Performance regression check: should be reasonably fast
        assert!(avg_ms < 100.0, "Parsing too slow: {:.2} ms", avg_ms);
    }
}

File: asterix-node/test/integration.test.js
/**
 * Integration tests for ASTERIX decoder Node.js bindings
 *
 * These tests verify end-to-end functionality with real ASTERIX data files
 * from the install/share/asterix/samples/ directory.
 *
 * Requirements:
 * - ASTERIX core library must be built and installed
 * - Sample files must be present in install/share/asterix/samples/
 */

const { expect } = require('chai');
const asterix = require('../lib/index');
const fs = require('fs');
const path = require('path');

describe('ASTERIX Decoder - Integration Tests', function() {
  // Increase timeout for file I/O and parsing
  this.timeout(10000);

  const samplesDir = path.join(__dirname, '../../install/share/asterix/samples');
  const configDir = path.join(__dirname, '../../install/share/asterix/config');

  before(function() {
    // Initialize parser with default config
    try {
      asterix.init(configDir);
      console.log(`Initialized ASTERIX parser with config: ${configDir}`);
    } catch (err) {
      console.warn('Initialization warning:', err.message);
      // Try auto-initialization
      try {
        asterix.init();
      } catch (err2) {
        console.error('Failed to initialize:', err2.message);
      }
    }
  });

  describe('Raw ASTERIX Files', function() {
    it('should parse cat048.raw', function() {
      const filePath = path.join(samplesDir, 'cat048.raw');

      // Skip if file doesn't exist
      if (!fs.existsSync(filePath)) {
        this.skip();
        return;
      }

      const data = fs.readFileSync(filePath);
      expect(data.length).to.be.greaterThan(0);

      const records = asterix.parse(data, { verbose: true });

      expect(records).to.be.an('array');
      expect(records.length).to.be.greaterThan(0);

      // Verify all records are CAT048
      records.forEach(record => {
        expect(record).to.have.property('category');
        expect(record.category).to.equal(48);
        expect(record).to.have.property('items');
        expect(record.items).to.be.an('object');
      });

      console.log(`  ✓ Parsed ${records.length} CAT048 records from cat048.raw`);
    });

    it('should parse cat034.raw', function() {
      const filePath = path.join(samplesDir, 'cat034.raw');

      if (!fs.existsSync(filePath)) {
        this.skip();
        return;
      }

      const data = fs.readFileSync(filePath);
      const records = asterix.parse(data);

      expect(records).to.be.an('array');
      expect(records.length).to.be.greaterThan(0);

      records.forEach(record => {
        expect(record.category).to.equal(34);
      });

      console.log(`  ✓ Parsed ${records.length} CAT034 records from cat034.raw`);
    });

    it('should parse cat062cat065.raw (multi-category)', function() {
      const filePath = path.join(samplesDir, 'cat062cat065.raw');

      if (!fs.existsSync(filePath)) {
        this.skip();
        return;
      }

      const data = fs.readFileSync(filePath);
      const records = asterix.parse(data);

      expect(records).to.be.an('array');
      expect(records.length).to.be.greaterThan(0);

      // Should contain both CAT062 and CAT065
      const categories = [...new Set(records.map(r => r.category))];
      expect(categories).to.include.members([62, 65]);

      console.log(`  ✓ Parsed ${records.length} records (CAT ${categories.join(', ')}) from cat062cat065.raw`);
    });
  });

  describe('PCAP Files', function() {
    it('should parse cat_062_065.pcap', function() {
      const filePath = path.join(samplesDir, 'cat_062_065.pcap');

      if (!fs.existsSync(filePath)) {
        this.skip();
        return;
      }

      const data = fs.readFileSync(filePath);

      // Parse PCAP format - may need special handling
      try {
        const records = asterix.parse(data);
        expect(records).to.be.an('array');

        if (records.length > 0) {
          console.log(`  ✓ Parsed ${records.length} records from cat_062_065.pcap`);
        } else {
          console.log('  ⚠ No records parsed (PCAP decapsulation may require special handling)');
        }
      } catch (err) {
        // PCAP may require different parsing approach
        console.log(`  ⚠ PCAP parsing not supported in Node.js binding: ${err.message}`);
        this.skip();
      }
    });

    it('should parse cat_034_048.pcap', function() {
      const filePath = path.join(samplesDir, 'cat_034_048.pcap');

      if (!fs.existsSync(filePath)) {
        this.skip();
        return;
      }

      const data = fs.readFileSync(filePath);

      try {
        const records = asterix.parse(data);
        expect(records).to.be.an('array');

        if (records.length > 0) {
          const categories = [...new Set(records.map(r => r.category))];
          console.log(`  ✓ Parsed ${records.length} records (CAT ${categories.join(', ')}) from cat_034_048.pcap`);
        }
      } catch (err) {
        console.log(`  ⚠ PCAP parsing not supported: ${err.message}`);
        this.skip();
      }
    });

    it('should parse asterix.pcap', function() {
      const filePath = path.join(samplesDir, 'asterix.pcap');

      if (!fs.existsSync(filePath)) {
        this.skip();
        return;
      }

      const data = fs.readFileSync(filePath);

      try {
        const records = asterix.parse(data);
        expect(records).to.be.an('array');

        if (records.length > 0) {
          const categories = [...new Set(records.map(r => r.category))];
          console.log(`  ✓ Parsed ${records.length} records (CAT ${categories.join(', ')}) from asterix.pcap`);
        }
      } catch (err) {
        console.log(`  ⚠ PCAP parsing not supported: ${err.message}`);
        this.skip();
      }
    });
  });

  describe('Incremental Parsing', function() {
    it('should parse large file incrementally without memory issues', function() {
      const filePath = path.join(samplesDir, 'cat048.raw');

      if (!fs.existsSync(filePath)) {
        this.skip();
        return;
      }

      const data = fs.readFileSync(filePath);
      let offset = 0;
      const allRecords = [];
      const chunkSize = 10; // Parse 10 blocks at a time

      while (offset < data.length) {
        const result = asterix.parseWithOffset(data, offset, chunkSize);

        expect(result).to.have.property('records');
        expect(result).to.have.property('bytesConsumed');
        expect(result).to.have.property('remainingBlocks');

        allRecords.push(...result.records);
        offset = result.bytesConsumed;

        if (result.remainingBlocks === 0) {
          break;
        }
      }

      expect(allRecords.length).to.be.greaterThan(0);
      console.log(`  ✓ Incrementally parsed ${allRecords.length} records in ${Math.ceil(allRecords.length / chunkSize)} chunks`);
    });

    it('should handle offset at exact block boundaries', function() {
      const filePath = path.join(samplesDir, 'cat062cat065.raw');

      if (!fs.existsSync(filePath)) {
        this.skip();
        return;
      }

      const data = fs.readFileSync(filePath);

      // Parse first block
      const result1 = asterix.parseWithOffset(data, 0, 1);
      expect(result1.bytesConsumed).to.be.greaterThan(0);

      // Parse second block starting exactly at end of first
      if (result1.bytesConsumed < data.length) {
        const result2 = asterix.parseWithOffset(data, result1.bytesConsumed, 1);
        expect(result2.records).to.be.an('array');
      }

      console.log('  ✓ Correctly handled block boundaries');
    });
  });

  describe('Category Filtering', function() {
    it('should filter by category correctly', function() {
      const filePath = path.join(samplesDir, 'cat062cat065.raw');

      if (!fs.existsSync(filePath)) {
        this.skip();
        return;
      }

      const data = fs.readFileSync(filePath);

      // Parse without filter
      const allRecords = asterix.parse(data);

      // Parse with CAT062 filter
      const cat62Records = asterix.parse(data, { filterCategory: 62 });

      // Parse with CAT065 filter
      const cat65Records = asterix.parse(data, { filterCategory: 65 });

      // Verify filtering worked
      cat62Records.forEach(r => expect(r.category).to.equal(62));
      cat65Records.forEach(r => expect(r.category).to.equal(65));

      // Filtered results should sum to total (approximately)
      expect(cat62Records.length + cat65Records.length).to.be.closeTo(allRecords.length, 5);

      console.log(`  ✓ Filtered CAT062: ${cat62Records.length}, CAT065: ${cat65Records.length}, Total: ${allRecords.length}`);
    });
  });

  describe('Max Records Limit', function() {
    it('should respect maxRecords option', function() {
      const filePath = path.join(samplesDir, 'cat048.raw');

      if (!fs.existsSync(filePath)) {
        this.skip();
        return;
      }

      const data = fs.readFileSync(filePath);

      // Parse with limit
      const maxLimit = 5;
      const records = asterix.parse(data, { maxRecords: maxLimit });

      expect(records.length).to.be.at.most(maxLimit);
      console.log(`  ✓ Limited to ${records.length} records (max: ${maxLimit})`);
    });
  });

  describe('Verbose Mode', function() {
    it('should include descriptions when verbose=true', function() {
      const filePath = path.join(samplesDir, 'cat048.raw');

      if (!fs.existsSync(filePath)) {
        this.skip();
        return;
      }

      const data = fs.readFileSync(filePath);

      // Parse with verbose mode
      const records = asterix.parse(data, { verbose: true, maxRecords: 1 });

      if (records.length > 0) {
        const record = records[0];

        // Should have items with descriptions
        expect(record.items).to.be.an('object');

        // At least one item should have a description
        const items = Object.values(record.items);
        const hasDescription = items.some(item =>
          item && typeof item === 'object' && 'description' in item
        );

        if (hasDescription) {
          console.log('  ✓ Verbose mode includes item descriptions');
        } else {
          console.log('  ⚠ No descriptions found (may not be available for all items)');
        }
      }
    });
  });

  describe('Category Information', function() {
    it('should correctly identify defined categories', function() {
      // Common ASTERIX categories
      const commonCategories = [1, 2, 8, 10, 19, 20, 21, 34, 48, 62, 65];

      const definedCategories = commonCategories.filter(cat =>
        asterix.isCategoryDefined(cat)
      );

      expect(definedCategories.length).to.be.greaterThan(0);
      console.log(`  ✓ Found ${definedCategories.length} defined categories: ${definedCategories.join(', ')}`);
    });

    it('should provide category descriptions', function() {
      const categories = [1, 2, 8, 10, 19, 20, 21, 34, 48, 62, 65];

      for (const cat of categories) {
        if (asterix.isCategoryDefined(cat)) {
          try {
            const desc = asterix.describe(cat);
            expect(desc).to.be.a('string');
            expect(desc.length).to.be.greaterThan(0);
            console.log(`  ✓ CAT${cat.toString().padStart(3, '0')}: ${desc.substring(0, 60)}...`);
          } catch (err) {
            // May not have description
          }
        }
      }
    });
  });

  describe('Performance', function() {
    it('should parse large files efficiently', function() {
      const filePath = path.join(samplesDir, 'cat048.raw');

      if (!fs.existsSync(filePath)) {
        this.skip();
        return;
      }

      const data = fs.readFileSync(filePath);

      // Warm-up
      asterix.parse(data);

      // Measure performance
      const iterations = 100;
      const start = Date.now();

      for (let i = 0; i < iterations; i++) {
        asterix.parse(data);
      }

      const elapsed = Date.now() - start;
      const avgTime = elapsed / iterations;
      const throughput = (data.length * iterations / 1024 / 1024) / (elapsed / 1000);

      console.log(`  ✓ Average parse time: ${avgTime.toFixed(2)}ms`);
      console.log(`  ✓ Throughput: ${throughput.toFixed(2)} MB/s`);

      // Performance target: should parse at least 1 MB/s
      expect(throughput).to.be.greaterThan(1);
    });
  });

  describe('Error Handling', function() {
    it('should handle invalid ASTERIX data gracefully', function() {
      const invalidData = Buffer.from([
        0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF
      ]);

      expect(() => asterix.parse(invalidData)).to.throw();
    });

    it('should handle truncated data gracefully', function() {
      const filePath = path.join(samplesDir, 'cat048.raw');

      if (!fs.existsSync(filePath)) {
        this.skip();
        return;
      }

      const data = fs.readFileSync(filePath);

      // Truncate to half length
      const truncated = data.subarray(0, Math.floor(data.length / 2));

      // Should either parse partial data or throw error
      try {
        const records = asterix.parse(truncated);
        expect(records).to.be.an('array');
        console.log(`  ✓ Parsed ${records.length} records from truncated data`);
      } catch (err) {
        expect(err).to.be.an('error');
        console.log(`  ✓ Correctly rejected truncated data: ${err.message}`);
      }
    });
  });
});
